File: 97-interleaving-string/97-interleaving-string.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        
         
        int m = s1.length();
        int n = s2.length();
        int N = s3.length();
        if(m + n != N)
            return false;
        
        bool t[n+1];
        
        
        for(int i = 0; i<m+1; i++) {
            for(int j = 0; j<n+1; j++) {
                if(i == 0 && j == 0) {
                    t[j] = true;
                } else if(i == 0) {
                    t[j] = t[j-1] && s2[j-1]==s3[i+j-1];
                } else if (j == 0) {
                    t[j] = t[j] && s1[i-1]==s3[i+j-1];
                } else {
                    t[j] = (t[j-1] && s2[j-1]==s3[i+j-1]) || (t[j] && s1[i-1]==s3[i+j-1]);
                }
            }
        }
        
        return t[n];
    }
        
        
        
    
};

File: 97-interleaving-string/97-interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1)+len(s2) !=len(s3):
            return False 
    

        q = deque()
        q.append((0,0))
        visit = set((0,0))
        while q: 
            x,y = q.popleft() 

            if x+y == len(s3):
                return True 


            if x+1 <=len(s1) and s1[x] ==s3[x+y] and (x+1,y) not in visit: 
                q.append((x+1,y)) 
                visit.add((x+1,y))
            if y+1 <=len(s2) and s2[y] ==s3[x+y] and (x,y+1) not in visit: 
                q.append((x,y+1))
                visit.add((x,y+1))

        return False 
